### ISSUE TITLE
How to incorporate additional information into a custom objective in xgboost?

### ISSUE THREAD

**Issue by rohitpatra** [dmlc/xgboost#7796]
Hi All, 

I am new to XgBoost and I am trying to create a custom objective, where the loss function depends on information other than the prediction and label. The only way to modify the hessian and the gradient for the XgBoost seems to be via get_weight(). 
For example as suggested [here](https://github.com/dmlc/xgboost/issues/3187#issuecomment-388026837)

def obj_rmse(preds, dtrain):
x = preds - dtrain.get_label()
grad = x
hess = np.ones(len(grad))
weights = dtrain.get_weight()
grad *= weights
hess *= weights
return grad, hess

I would like to use weights to pass on other information to compute the loss. However, I am worried if ".get_weight()" is used by XGBoost at **other places** during the training. 
If yes, then I might be making some other changes to the training. In that case would  anyone be able to suggest how to pass more information in to the custom objective. 

Thank you for your time and help.

Best,
Rohit

**Comment by trivialfis** [comment]
Consider the following examples:
``` python
my_other_info = np.array(...)

def my_obj(predt, dtrain):
    y = dtrain.get_label()
    grad = gradient(y, my_other_info, predt)
    hess = hessian(y, my_other_info, predt)
    ...
```

``` python
class MyObj:
    def __init__(self):
        self.my_other_info = np.array(...)

    def __call__(self, predt, dtrain):
        grad = gradient(dtrain.get_label(), self.my_other_info, predt)
        ...
```

**Comment by trivialfis** [comment]
> Is dtrain identical across all the calls to the custom objective function? And if yes, does every call to the custom objective use the whole data (or is there some sampling mechanism)?

Yes. The whole data, no sampling occurs in the objective.

> However if dtrain is different across different calls then I would not be able to use my_other_info as a single constant array.

As above.

> Sorry for more questions. Thank you for your help.

You are welcome.
